### ISSUE TITLE
feat(merge): DAM — Differentiable Adaptive Merging with trainable coefficients

### ISSUE THREAD

**Issue by noahgift** [paiml/aprender#446]
## Summary

Implement DAM (Differentiable Adaptive Merging): trainable per-column merge
coefficients optimized via gradient descent on a calibration dataset. Instead
of fixed merge weights, DAM learns the optimal per-parameter merge strategy.

## CLI

```bash
apr merge dam \
  model-a.apr model-b.apr \
  --base-model base.apr \
  --data calibration.jsonl \
  --dam-epochs 5 \
  --dam-lr 0.01 \
  -o merged.apr
```

## Parameters

- `--data`: Calibration dataset for coefficient optimization
- `--dam-epochs`: Training epochs for merge coefficients (default: 5)
- `--dam-lr`: Learning rate for coefficient optimization (default: 0.01)

## Implementation

1. Initialize per-column merge coefficients (uniform)
2. Forward pass: merge models using current coefficients
3. Compute loss on calibration data
4. Backprop through merge to update coefficients
5. Iterate for N epochs
6. Apply learned coefficients for final merge

## Reference

[DAM GitHub](https://github.com/arcee-ai/DAM) — open source by Arcee.

## Specification

See `docs/specifications/components/merge.md` §4.10.

### Contract: `dam-merge-v1.yaml`

```yaml
metadata:
  description: "Differentiable Adaptive Merging — trainable per-column merge coefficients"
  references:
    - "Arcee AI (2024) DAM: Differentiable Adaptive Merging"
  depends_on:
    - "lora-algebra-v1"
    - "adamw-kernel-v1"

equations:
  dam_merge:
    formula: "θ_merged(j) = Σ α_i(j) * θ_i(j), where α learned via gradient descent"
    invariants:
      - "Σ α_i(j) = 1 for each column j (softmax normalization)"
      - "α_i(j) ∈ (0, 1) after softmax"
      - "shape(θ_merged) == shape(θ_base)"

  coefficient_update:
    formula: "α ← α - η * ∂L/∂α (gradient descent on calibration loss)"
    invariants:
      - "Loss decreases monotonically (with sufficiently small η)"
      - "Coefficients remain valid after update (softmax maintains simplex)"

proof_obligations:
  - type: invariant
    property: "Coefficient normalization"
    formal: "sum(softmax(raw_coeffs[j])) == 1.0 for all columns j"
  - type: invariant
    property: "Shape preservation"
    formal: "shape(dam_merge(models, coeffs)) == shape(models[0])"
  - type: monotonicity
    property: "Training loss decrease"
    formal: "L(α_{t+1}) <= L(α_t) for small η"

falsification_tests:
  - id: FALSIFY-DAM-001
    rule: "Coefficient normalization"
    prediction: "sum(coefficients) == 1.0 per column after softmax"
    if_fails: "Softmax normalization not applied to coefficients"
  - id: FALSIFY-DAM-002
    rule: "Uniform init equivalence"
    prediction: "DAM with uniform coefficients == simple average merge"
    if_fails: "Initialization or merge formula incorrect"
  - id: FALSIFY-DAM-003
    rule: "Gradient flow"
    prediction: "∂L/∂α is non-zero for non-trivial calibration data"
    if_fails: "Gradient computation through merge is broken"
```

### Binding Requirements

```yaml
  - contract: dam-merge-v1.yaml
    equation: dam_merge
    module_path: "aprender::format::apr_merge"
    function: merge_dam
    status: not_implemented
```

### Verification Targets
- L2: 3 falsification tests
- L3: Kani harness for coefficient normalization (bound dim≤4)

**Comment by noahgift** [comment]
Implemented. DAM (Differentiable Adaptive Merging) in `src/online/dam.rs` (385 lines) + `src/online/dam_tests.rs`. Trainable per-column merge coefficients optimized via gradient descent on calibration data.
